pkg/verifier/notation_verifier.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/notaryproject/notation-core-go/signature"
+	// Blank import only needed to trigger init() and have "application/cose"
+	// added to corresponding map:
+	// https://github.com/notaryproject/notation-core-go/blob/a9c5e3f1a5b5/signature/cose/envelope.go#L46
+	_ "github.com/notaryproject/notation-core-go/signature/cose"
+)
+
+type NotationVerifyOptions struct {
+	Certificates []string
+}
+
+type notationVerifier struct {
+	certificates []string
+}
+
+func getSignatures(ctx context.Context, repo *remote.Repository, imageStore oras.Target, ref reference.Named) ([][]byte, error) {
+	desc, err := imageStore.Resolve(ctx, ref.String())
+	if err != nil {
+		return nil, fmt.Errorf("resolving image %q: %w", ref, err)
+	}
+
+	signaturesBytes := make([][]byte, 0)
+
+	err = repo.Referrers(ctx, desc, "application/vnd.cncf.notary.signature", func(references []ocispec.Descriptor) error {
+		for _, reference := range references {
+			rc, err := repo.Fetch(ctx, reference)
+			if err != nil {
+				return fmt.Errorf("fetching signature manifest content: %w", err)
+			}
+			defer rc.Close()
+
+			bytes, err := io.ReadAll(rc)
+			if err != nil {
+				return fmt.Errorf("getting signature manifest bytes: %w", err)
+			}
+
+			signatureManifest := &ocispec.Manifest{}
+			err = json.Unmarshal(bytes, signatureManifest)
+			if err != nil {
+				return fmt.Errorf("decoding signature manifest: %w", err)
+			}
+
+			for _, layer := range signatureManifest.Layers {
+				// JWS are not supported.
+				if layer.MediaType != "application/cose" {
+					continue
+				}
+
+				rc, err := repo.Fetch(ctx, layer)
+				if err != nil {
+					return fmt.Errorf("fetching signature content: %w", err)
+				}
+
+				signatureBytes, err := io.ReadAll(rc)
+				rc.Close()
+				if err != nil {
+					return fmt.Errorf("getting signature bytes: %w", err)
+				}
+
+				signaturesBytes = append(signaturesBytes, signatureBytes)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("ranging through referrers: %w", err)
+	}
+
+	return signaturesBytes, nil
+}
+
+// Inspired by:
+// https://github.com/notaryproject/notation-go/blob/6063ebe30f96b0830e1385db4cdcdc24b0536d1a/verifier/verifier.go#L681
+func getEnveloppe(signatureBytes []byte) (*signature.EnvelopeContent, error) {
+	enveloppe, err := signature.ParseEnvelope("application/cose", signatureBytes)
+	if err != nil {
+		return nil, fmt.Errorf("parsing envelope: %w", err)
+	}
+
+	content, err := enveloppe.Verify()
+	if err != nil {
+		return nil, fmt.Errorf("verifying envelope: %w", err)
+	}
+
+	if content.Payload.ContentType != "application/vnd.cncf.notary.payload.v1+json" {
+		return nil, fmt.Errorf("bad payload content type, expected \"application/vnd.cncf.notary.payload.v1+json\", got: %q", content.Payload.ContentType)
+	}
+
+	return content, nil
+}
+
+func (n *notationVerifier) Verify(ctx context.Context, repo *remote.Repository, imageStore oras.Target, ref reference.Named) error {
```

Comment:
It is more suggested and securer to call APIs of `notation-go` instead of your own logic from `notation-core-go` unless you know what you are doing.